pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/utils/RealTimeRegexpMatcher.java

```diff
@@ -0,0 +1,156 @@
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.utils.nativefst.utils;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.Automaton;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.CharacterRunAutomaton;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.RegExp;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.State;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.Transition;
+import org.apache.pinot.segment.local.utils.nativefst.mutablefst.MutableArc;
+import org.apache.pinot.segment.local.utils.nativefst.mutablefst.MutableFST;
+import org.apache.pinot.segment.local.utils.nativefst.mutablefst.MutableState;
+import org.roaringbitmap.IntConsumer;
+
+public class RealTimeRegexpMatcher {
+  private final String _regexQuery;
+  private final MutableFST _fst;
+  private final Automaton _automaton;
+  private final IntConsumer _dest;
+
+  public RealTimeRegexpMatcher(String regexQuery, MutableFST fst, IntConsumer dest) {
+    _regexQuery = regexQuery;
+    _fst = fst;
+    _dest = dest;
+
+    _automaton = new RegExp(_regexQuery).toAutomaton();
+  }
+
+  public static void regexMatch(String regexQuery, MutableFST fst, IntConsumer dest) {
+    RealTimeRegexpMatcher matcher = new RealTimeRegexpMatcher(regexQuery, fst, dest);
+    matcher.regexMatchOnFST();
+  }
+
+  // Matches "input" string with _regexQuery Automaton.
+  public boolean match(String input) {
+    CharacterRunAutomaton characterRunAutomaton = new CharacterRunAutomaton(_automaton);
+    return characterRunAutomaton.run(input);
+  }
+
+  /**
+   * This function runs matching on automaton built from regexQuery and the FST.
+   * FST stores key (string) to a value (Long). Both are state machines and state transition is based on
+   * a input character.
+   *
+   * This algorithm starts with Queue containing (Automaton Start Node, FST Start Node).
+   * Each step an entry is popped from the queue:
+   *    1) if the automaton state is accept and the FST Node is final (i.e. end node) then the value stored for that FST
+   *       is added to the set of result.
+   *    2) Else next set of transitions on automaton are gathered and for each transition target node for that character
+   *       is figured out in FST Node, resulting pair of (automaton state, fst node) are added to the queue.
+   *    3) This process is bound to complete since we are making progression on the FST (which is a DAG) towards final
+   *       nodes.
+   */
+  public void regexMatchOnFST() {
```

Comment:
remove 9 blank lines

Reply:
For the blank lines, may I suggest that while these are a bit subjective, there are some practical implications of these:
- Readability of code on whether it looks like a giant wall of text, vs too sparse such that a method won't fit in a page.
- Next person touching the code may alter these, leading to unnecessary diffs.

My $0.02 would be to enforce this via formatter, or guidelines (if formatter rules cannot be customized).

Reply:
So in this scenario, we are at an impasse because we neither have formatter rules nor guidelines. I will concede and remove all blank lines, per Richard's command.

Reply:
@mayankshriv the formatter leaves it up to the author for a few reasons:

1. it's difficult to make the tools enforce a guideline like this only within method bodies - the only tool available is RegexpMultiline which is a blunt instrument
2. it's expensive (in terms of build time) to apply the numerous RegexpMultiline rules necessary
3. A lot of the time, a line _here and there_ makes code more readable (though it usually indicates that the enclosing scope is too large and needs to be broken down), but this PR consistently takes this grey area to the extreme. For instance, consider the constructor below: what do the blank lines mean? Are they intentional? Why is initialising `_pathState` different to `_state`, `_node`, or `_fstArc`? As a peer reviewing this change set, I think it's inconsistent and a waste of screen space.

```java
    public Path(State state, MutableState node, MutableArc fstArc, List<Character> pathState) {
       _state = state;
       _node = node;
       _fstArc = fstArc;

       _pathState = pathState;

       _pathState.add(node.getLabel());
     }
```

Reply:
@richardstartin First off, I really appreciate your in-depth look into the PR. As usual, your comments are invaluable.
I agree with you that the spacing on the PR was a bit too much -- and I have fixed it per your comments. However, I would also ask yourself to be a bit considerate for expression of a personal style. The example you made above about inconsistent spacing makes perfect space, and I am happy to fix that (and would have fixed it in a blink, had it been the primary comment that you made about spacing). However, you have commented on nearly every blank line that I added in this PR, including all the places where I was consistent. For eg., I added a blank space after variable declaration and before usage, and this is a consistent pattern. You did not agree with the same and asked me to revert that in mostly every file that this PR consists of. I accepted your comments and changed the same. All I am requesting you (humbly) is to allow me a bit of room (within reasonable limits, of course). I tend to find adding a blank line soothing to my eyes, and while I understand that this PR went overboard with the same, I have tried my best to fix it now. I request you to kindly not allege that I am being intentional by adding too many lines and trying to obfuscate the project (quoting your latest comment). I sincerely apologise if I have crossed any boundaries or have been detrimental to the project in any form -- that is not my intention.

And, as you must have noticed, I have followed your advise to not add spaces in my text anymore. This comment is a testament to that.

Reply:
> However, I would also ask yourself to be a bit considerate for expression of a personal style. 
...
> All I am requesting you (humbly) is to allow me a bit of room (within reasonable limits, of course)

I don't believe there is room for personal expression in a shared codebase - it's engineering, not poetry. Ideally it shouldn't be possible to tell who wrote the code without looking at git blame. More time will be spent reading this code than was spent writing it, so it all needs to look the same for the sake of the person who may need to make changes in the future.

> I request you to kindly not allege that I am being intentional by adding too many lines and trying to obfuscate the project (quoting your latest comment)

I don't know what you are referring to - no allegations have been made about intent to obfuscate.

Reply:
Sorry for jumping in the discussion late. I myself have the habit to add many line separators for readability as well so glad to see some discussions about this. In general I found it hard to enforce a checkstyle on readability as I agree it is rather subjective. (other than those obvious ones such as https://google.github.io/styleguide/javaguide.html). 

my own opinion here is to follow as close to the OSS codebase standard from as possible - this gives a consistency sense to users as well as first-time contributors. But without incurring too much overhead, I don't think we want to discourage contribution because of formatting in general.

but i saw richard here provided many constructive feedback as well beside the formatting; in my experience ppl who reviewed extensively on such a large PR will also be the ones who maintain it and collaborate together going forward. it would be great to reach a reasonable readability middle ground.


This is just my personal opinion, I am also pretty new to Pinot so please disregard if it doesn't seem to make sense. But definitely will use this as a standard for my contribution going forward.

Reply:
Sorry to resurrect this thread but none of the underlying issues (profuse blank lines) have been addressed in this class.

Reply:
Fixed to best of my knowledge